netutil/netlink.go

```diff
@@ -0,0 +1,122 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package netutil
+
+import (
+	"net"
+	"syscall"
+)
+
+const (
+	// see /usr/include/linux/rtnetlink.h
+	RTMGRP_IPV4_ROUTE = 0x40
+	RTMGRP_IPV6_ROUTE = 0x400
+
+	RTM_NEWROUTE = 24
+	RTM_DELROUTE = 25
+)
+
+type RoutesMonitor struct {
+	netlinkFd int
+
+	defaultGwAdded   func(string)
+	defaultGwRemoved func(string)
+}
+
+func NewRoutesMonitor(defaultGwAdded, defaultGwRemoved func(string)) *RoutesMonitor {
+	m := &RoutesMonitor{
+		defaultGwAdded:   defaultGwAdded,
+		defaultGwRemoved: defaultGwRemoved,
+	}
+	return m
+}
+
+func (m *RoutesMonitor) Connect() error {
+	fd, err := syscall.Socket(syscall.AF_NETLINK, syscall.SOCK_RAW, syscall.NETLINK_ROUTE)
+	if err != nil {
+		return err
+	}
+	addr := &syscall.SockaddrNetlink{
+		Family: syscall.AF_NETLINK,
+		Groups: RTMGRP_IPV4_ROUTE | RTMGRP_IPV6_ROUTE,
+	}
+	if err := syscall.Bind(fd, addr); err != nil {
+		return err
+	}
+	m.netlinkFd = fd
+	go m.run()
+	return nil
+}
+
+func (m *RoutesMonitor) Disconnect() {
+	syscall.Close(m.netlinkFd)
+	m.netlinkFd = 0
+}
+
+func isDefaultGw(mm *syscall.NetlinkMessage) (bool, net.IP) {
+	nras, err := syscall.ParseNetlinkRouteAttr(mm)
+	if err != nil {
+		// XXX: log error?
+		return false, nil
+	}
+	for _, nra := range nras {
+		switch nra.Attr.Type {
+		case syscall.RTA_GATEWAY:
+			return true, net.IP(nra.Value)
+		}
+	}
+	return false, nil
+}
+
+func (m *RoutesMonitor) run() {
+	buf := make([]byte, syscall.Getpagesize())
+	for {
```

Comment:
when does this stop?  I think for ease of reasoning we should make the interface of this monitor as close as possible to udevmonitor.Monitor

Reply:
It will stop when the netlinkFd is closed, but I'm happy to change this to be closer to the udevmontior.Monitor.

Reply:
? I don't see a break or a return here, what am I missing? anyway I'm happy also if you simplify udevmonitor if that make sense, but yes ideally their surface should be similar

Reply:
I'm in agreement with Samuele. Whatever the best way of doing things we should do it and be consistent.

Reply:
Uh, silly me - I spoke too early. If the general feature looks reasonable I'm happy to improve this and also write a better test, I will try to simulate netlink traffic in a unittest to ensure this can actually stop.